### PR TITLE
[StatePrepAliasSampling] Fix expression for sampling bitsize `mu`

### LINKED ISSUE #1071: Bitlength of "keep" and "alt" probabilities in coherent alias sampling
Qualtran computes this as log_2(ε/n):
https://github.com/quantumlib/Qualtran/blob/0d8793e8a7f415d6ec1e003756422d72ac81a3a1/qualtran/linalg/lcu_util.py#L161
and does not make any reference to λ (the 1-norm / subnormalisation).

However, this doesn't match the papers:
*  Eq. (36) in "Encoding Electronic Spectra in Quantum Circuits with Linear T Complexity" has a different formula with dependence on λ.
* Eq. (A12) in the Tensor Hypercontraction paper also has a formula for the bit length of the keep probabilities that makes reference to λ.

Tagging @tanujkhattar whom I briefly discussed this inconsistency with at the Qualtran workshop last Friday

### PR DESCRIPTION
**Pull request by anurudhp**
fixes #1071 

- compute the precision for prepare from the required precision in eigenphases.
- weak norm upper bounds for ising/pauli hamiltonians
- rename `lcu_coeff` -> `probabilities` (as the modules are more generally applicable)

### CODE REVIEW

**Comment by anurudhp** [comment]
@tanujkhattar ptal!

File: qualtran/bloqs/chemistry/ising/walk_operator.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    for i in range(L):
+        for j in range(i):
+            if not cirq.commutes(ham_ps[i], ham_ps[j]):
+                anticommute_sum += np.abs(ham_ps[i].coefficient * ham_ps[j].coefficient)
+
+    return np.sqrt(qlambda**2 - 2 * anticommute_sum)
+
+
+def walk_operator_for_pauli_hamiltonian(
+    ham: cirq.PauliSum, eps: float, *, ham_norm_upper_bound: Optional[float] = None
```
Comment (tanujkhattar) [inline qualtran/bloqs/chemistry/ising/walk_operator.py:105]: Should `eps` be a `SymbolicFloat` ?

File: qualtran/bloqs/reflections/reflection_using_prepare_test.py [hunk 2]
```diff
@@ -223,8 +232,8 @@ def test_reflection_using_prepare_diagram():
 
 
 def test_reflection_using_prepare_consistent_protocols_and_controlled():
-    prepare_gate = StatePreparationAliasSampling.from_lcu_probs(
-        [1, 2, 3, 4], probability_epsilon=0.1
+    prepare_gate = StatePreparationAliasSampling.from_probabilities(
+        [1, 2, 3, 4], precision=0.1 * 10
     )
```
Comment (tanujkhattar) [inline qualtran/bloqs/reflections/reflection_using_prepare_test.py:237]: This change is unrelated to the PR now ?

File: qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py [hunk 3]
```diff
@@ -96,66 +96,67 @@ class StatePreparationAliasSampling(PrepareOracle):
...
     """
+
     selection_registers: Tuple[Register, ...] = attrs.field(
         converter=lambda v: (v,) if isinstance(v, Register) else tuple(v)
     )
     alt: Union[Shaped, NDArray[np.int_]]
     keep: Union[Shaped, NDArray[np.int_]]
     mu: SymbolicInt
-    sum_of_lcu_coeffs: SymbolicFloat
+    sum_of_coefficients: SymbolicFloat
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py:107]: Also, I think `sum_of_coefficients` is not actually used by this class but I remember it was added so `QubitizationWalkOperator` can make use of it. @charlesyuan314 - it's worth taking a look as part of your design and see if we really should have the lambda exposed as a property from the state preparation bloqs and if yes, let's refactor `StatePreparationViaRotations` to also do the same.

File: qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py [hunk 4]
```diff
@@ -96,66 +97,76 @@ class StatePreparationAliasSampling(PrepareOracle):
...
         )
 
     @classmethod
     def from_n_coeff(
-        cls,
-        n_coeff: SymbolicInt,
-        sum_of_lcu_coeffs: SymbolicFloat,
-        *,
-        probability_epsilon: SymbolicFloat = 1.0e-5,
+        cls, n_coeff: SymbolicInt, sum_of_terms: SymbolicFloat, *, precision: SymbolicFloat = 1.0e-5
```
Comment (tanujkhattar) [inline qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py:147]: We should keep the naming consistent with the class attribute and factory method. The attribute has `sum_of_coefficients` and this has `sum_of_terms`.

File: qualtran/linalg/lcu_util.py [hunk 5]
```diff
@@ -148,69 +150,98 @@ def _preprocess_for_efficient_roulette_selection(discretized_probabilities):
...
+
+    Given unnormalized probabilities $w_l$ s.t. $\sum_l w_l = \lambda$, define
+    probabilities $p_l = w_l / \lambda$. This method computes the number of
+    bits $\mu$ to approximate the values $p_l$ such that the state prepared
+    by alias sampling has probabilities $\tilde{w_l} / \lambda$ satisfying
+    $\abs{w_l - \tilde{w}_l} \le \epsilon$.
+
+    Args:
+        number_of_coefficients: number of probabilities $L$.
+        precision: precision $|epsilon$ to approximate the unnormalized input
```
Comment (tanujkhattar) [inline qualtran/linalg/lcu_util.py:177]: `$|epsilon$` seems like a type? Maybe `$\epsilon$`

**Review by tanujkhattar — APPROVED** [review]
LGTM % comments.

File: qualtran/bloqs/qubitization/qubitization_walk_operator_test.py [hunk 6]
```diff
@@ -104,25 +88,24 @@ def test_qubitization_walk_operator(num_sites: int, eps: float):
         )
 
 
+@pytest.mark.slow()
 def test_qubitization_walk_operator_adjoint():
-    num_sites, eps = 4, 2e-1
```
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/bloqs/qubitization/qubitization_walk_operator_test.py:108]: Yeah, it was `0` before, and `3` now (after accounting for qlambda)

File: qualtran/linalg/lcu_util.py [hunk 7]
```diff
@@ -148,69 +150,98 @@ def _preprocess_for_efficient_roulette_selection(discretized_probabilities):
...
+def sub_bit_prec_from_epsilon(
+    number_of_coefficients: SymbolicInt, precision: SymbolicFloat
+) -> SymbolicInt:
+    r"""Number of bits to approximate the probabilities.
+
+    Given unnormalized probabilities $w_l$ s.t. $\sum_l w_l = \lambda$, define
+    probabilities $p_l = w_l / \lambda$. This method computes the number of
+    bits $\mu$ to approximate the values $p_l$ such that the state prepared
+    by alias sampling has probabilities $\tilde{w_l} / \lambda$ satisfying
+    $\abs{w_l - \tilde{w}_l} \le \epsilon$.
```
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/linalg/lcu_util.py]: oops, yeah you're right. fixed now

File: qualtran/linalg/lcu_util.py [hunk 8]
```diff
@@ -148,69 +150,98 @@ def _preprocess_for_efficient_roulette_selection(discretized_probabilities):
...
+
+    Args:
+        number_of_coefficients: number of probabilities $L$.
+        precision: precision $|epsilon$ to approximate the unnormalized input
+                   probabilities $w_l$ in alias sampling.
```
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/linalg/lcu_util.py]: fixed

File: qualtran/linalg/lcu_util.py (hunk 5 above)
Comment (anurudhp) [reply qualtran/linalg/lcu_util.py]: fixed

File: qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py [hunk 9]
```diff
@@ -96,66 +96,67 @@ class StatePreparationAliasSampling(PrepareOracle):
...
         )
-        N = len(lcu_probabilities)
+        N = len(unnormalized_probabilities)
         return StatePreparationAliasSampling(
             selection_registers=Register('selection', BoundedQUInt((N - 1).bit_length(), N)),
             alt=np.array(alt),
             keep=np.array(keep),
             mu=mu,
-            sum_of_lcu_coeffs=sum(abs(x) for x in lcu_probabilities),
+            sum_of_coefficients=sum(abs(x) for x in unnormalized_probabilities),
```
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py]: raised a value error in this method

File: qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py]: For now, I renamed it to `qlambda` and explained it in the docstring.

As for interface, `PrepareOracle` already has this property, so we should just override it in `StatePreparationViaRotations`.

File: qualtran/bloqs/chemistry/ising/walk_operator.py (hunk 1 above)
Comment (anurudhp) [reply qualtran/bloqs/chemistry/ising/walk_operator.py:105]: good point. The current impl doesn't support it, but it's a small change, let me fix that.

Comment (anurudhp) [reply qualtran/bloqs/chemistry/ising/walk_operator.py]: added support for symbolic eps and ham_norm. Making J, Gamma symbolic would require more work, as `cirq.PauliSum` does not support symbolic coeffs. Deferring to a later PR.

File: qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py (hunk 4 above)
Comment (anurudhp) [reply qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py]: fixed

**Review by tanujkhattar — APPROVED** [review]
